src/Microsoft.JSInterop/DotNetDispatcher.cs

```diff
@@ -0,0 +1,181 @@
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Provides methods that receive incoming calls from JS to .NET.
+    /// </summary>
+    public static class DotNetDispatcher
+    {
+        private static ConcurrentDictionary<string, IReadOnlyDictionary<string, (MethodInfo, Type[])>> _cachedMethodsByAssembly
+            = new ConcurrentDictionary<string, IReadOnlyDictionary<string, (MethodInfo, Type[])>>();
+
+        /// <summary>
+        /// Receives a call from JS to .NET, locating and invoking the specified method.
+        /// </summary>
+        /// <param name="assemblyName">The assembly containing the method to be invoked.</param>
+        /// <param name="methodIdentifier">The identifier of the method to be invoked. The method must be annotated with a <see cref="JSInvokableAttribute"/> matching this identifier string.</param>
+        /// <param name="argsJson">A JSON representation of the parameters.</param>
+        /// <returns>A JSON representation of the return value, or null.</returns>
+        public static string Invoke(string assemblyName, string methodIdentifier, string argsJson)
+        {
+            var syncResult = InvokeSynchronously(assemblyName, methodIdentifier, argsJson);
+            return syncResult == null ? null : Json.Serialize(syncResult);
+        }
+
+        /// <summary>
+        /// Receives a call from JS to .NET, locating and invoking the specified method asynchronously.
+        /// </summary>
+        /// <param name="callId">A value identifying the asynchronous call that should be passed back with the result, or null if no result notification is required.</param>
+        /// <param name="assemblyName">The assembly containing the method to be invoked.</param>
+        /// <param name="methodIdentifier">The identifier of the method to be invoked. The method must be annotated with a <see cref="JSInvokableAttribute"/> matching this identifier string.</param>
+        /// <param name="argsJson">A JSON representation of the parameters.</param>
+        /// <returns>A JSON representation of the return value, or null.</returns>
+        public static void BeginInvoke(string callId, string assemblyName, string methodIdentifier, string argsJson)
+        {
+            var syncResult = InvokeSynchronously(assemblyName, methodIdentifier, argsJson);
+
+            // If there was no callId, the caller does not want to be notified about the result
+            if (callId != null)
+            {
+                // Invoke and coerce the result to a Task so the caller can use the same async API
+                // for both synchronous and asynchronous methods
+                var task = syncResult is Task syncResultTask ? syncResultTask : Task.FromResult(syncResult);
+                task.ContinueWith(completedTask =>
```

Comment:
Do you anticipate any problems if the continuation executes synchronously? I believe the existing callers of this method currently behave correctly

Reply:
I don't think there will be any problem with that, because this method is `void`, so the caller can't be relying on receiving any synchronous output before the continuation is invoked.

Also since the other side of the interop boundary is getting the continuation as a `Promise` resolution, that already guarantees as per the `Promise` spec that resolutions will be delivered asynchronously. So in practice I don't think it will be possible to observe the callback arriving synchronously.

If anyone ever manages to create a scenario where the difference is observable, we can tweak this to force an async continuation.